### PR TITLE
Feature: Add cache to significantly speedup fetching from remote proxies

### PR DESCRIPTION
**Pull request by cardil**
I've added a cache that works in memory to significantly speedup subsequent remote fetches.

### CODE REVIEW

**Comment by cardil** [comment]
A bit strange. It works on my Ruby 1.9. I've double checked it

[asciicast](https://asciinema.org/a/32051)

**Comment by i11** [comment]
Hi,

Sorry for such late response. I got out of the loop totally.
Travis job still fails with the same thing. Not sure why. I'll be testing a bit more today.

**Comment by i11** [comment]
Btw, I saw you had a few more commits after this PR. Feel free to re-submit it.

**Comment by i11** [comment]
I wonder if it could be the bundle issue. 1.7.x seems like a pretty old version.
Could you please try updating .travis.yml with:
```
before_install:
  - gem update bundler
```

**Comment by i11** [comment]
Just tried and it does the trick.

File: lib/puppet_forge_server/backends/proxy.rb [hunk 1]
```diff
@@ -59,6 +63,27 @@ def upload(file_data)
...
+      path = Dir["#{@cache_dir}/**/#{file_name}"].first
+      unless File.exist?("#{path}")
+        buffer = get(relative_url)
+        File.open(target_file, 'wb') do |file|
+          file.write(buffer)
+          @log.debug("Saved #{buffer} bytes in filesystem cache for url: #{relative_url}, target file: #{target_file}")
+        end
+        path = target_file
+      else
+        @log.info("Filesystem cache HIT for url: #{relative_url}")
```
Comment (i11) [inline lib/puppet_forge_server/backends/proxy.rb]: I would put this one into debug

File: lib/puppet_forge_server/backends/proxy.rb [hunk 2]
```diff
@@ -35,14 +35,18 @@ def initialize(url, cache_dir, http_client, file_path)
...
-          file.write(buffer.read)
+        File.open(target_file, 'wb') do |file|
+          bytes = buffer.read
+          file.write(bytes)
+          @log.debug("Saved #{bytes.size} bytes in filesystem cache for path: #{relative_path}, target file: #{target_file}")
         end
-        path = File.join(@cache_dir, file_name[0].downcase, file_name)
+        path = target_file
+      else
+        @log.info("Filesystem cache HIT for path: #{relative_path}")
```
Comment (i11) [inline lib/puppet_forge_server/backends/proxy.rb]: I would put this one into debug

File: lib/puppet_forge_server/http/http_client.rb [hunk 3]
```diff
@@ -45,11 +54,29 @@ def download(url)
...
+      inspected.gsub(/>$/, ", @cache=#{cache_inspected}>")
+    end
+
     private
+
     def open_uri(url)
-      ::Timeout.timeout(10) do
-        open(url, 'User-Agent' => "Puppet-Forge-Server/#{PuppetForgeServer::VERSION}", :allow_redirections => :safe)
+      hit_or_miss = @cache.include?(url) ? 'HIT' : 'MISS'
+      @log.info "Cache in RAM memory size: #{@cache.size}, #{hit_or_miss} for url: #{url}"
```
Comment (i11) [inline lib/puppet_forge_server/http/http_client.rb]: I would put this one into debug

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -23,8 +24,47 @@
 module RSpecMixin
   include Rack::Test::Methods
 end
+begin
```
Comment (i11) [inline spec/spec_helper.rb]: Feels a bit out of context here...
Would it be more suitable to move some of this stuff into Rakefile instead?

File: lib/puppet_forge_server/backends/proxy.rb (hunk 1 above)
Comment (cardil) [inline lib/puppet_forge_server/backends/proxy.rb]: What do you mean?

This is main speedup factor - to cache non mutable file (metadata about specific version) on HDD

**Comment by cardil** [comment]
I would like it work without need for upgrading bundler.

```
before_install:
  - gem update bundler
```

I will look at it when I have a bit of time

**Comment by i11** [comment]
Oh and thanks a lot! This is great stuff.

**Comment by i11** [comment]
Sure thing

File: lib/puppet_forge_server/backends/proxy.rb (hunk 1 above)
Comment (i11) [inline lib/puppet_forge_server/backends/proxy.rb]: I mean change ```@log.info``` into ```@log.debug```

**Comment by i11** [comment]
I'm merging this. I'll add the gem update command for now as it doesn't go through for all PRs now.
Please feel very welcome to look into it when you have time.

**Comment by i11** [comment]
Quick though. It would be nice to have an endpoint for invalidating the cache.

**Comment by i11** [comment]
Now when I think about it caching all requests has heavy negative effects. For example after the module deployment TTL will be in the way for retrieving it. ```puppet module install`` will do ```/v3/releases?module=<module name>``` and it will be most probably already cached, so it won't find the new version. Hence search and release requests should have very short ttl or not cached at all. Already there things start to get complicated and squid or varnish feels much better choice.

**Comment by i11** [comment]
Module files on the other hand most probably will be sitting in RAM anyway as OS file cache, so benefits of explicit caching kind of disappears...

**Comment by cardil** [comment]
Endpoint for cache invalidation would be nice.

About caching: I have done in with two levels. First exist in ram and last by default 15min. So after release cached time is max 15min. I think it is acceptable. If you don't like this you have option to completely turn the cache down by settings either ttl or size to 0. Try to execute `librarian-puppet update` couple times with and without ram cache. The difference is huge. The bigger your Puppetfile gets the more benefits you would have from this cache.

And there is cache level 2 for requests that can't change. That is module tar.gz and meta information about specific version. I also ecurage to test with ram cache turned off and cleared persistent cache and not. In my case fetching single meta information about specific version takes about a second. I have about 30 dependencies in my Puppetfile so the difference for my is about 30 seconds. That is huge because I run it on every provisioning of Puppet master with vagrant. That's annoying, that slow down when run 30 times a day.

Better to give options to all of this so users can adjust puppet-forge-server to its needs.

**Comment by i11** [comment]
I fully understand yours and probably others need for caching. My point was that I'm not sure it should not be part of this project. Varnish or squid for example could provide all possible bells and whistles for whatever the needs are in that area. I don't think we should be reinventing the wheel here.

Anyway we took the pill already, let it be. But let's not extend it anymore. Also I'll be setting the default cache ttl to 0 in the next release, so that it  would be more explicit.

**Comment by cardil** [comment]
That seams completely reasonable.

**Pull request merged**